module/zfs/dmu_zfetch.c

```diff
@@ -407,20 +416,28 @@ dmu_zfetch_prepare(zfetch_t *zf, uint64_t blkid, uint64_t nblks,
 	 * Start prefetch from the demand access size (nblks).  Double the
 	 * distance every access up to zfetch_min_distance.  After that only
 	 * if needed increase the distance by 1/8 up to zfetch_max_distance.
+	 *
+	 * Don't double the distance beyond single block if we have more
+	 * than ~6% of ARC held by active prefetches.  It should help with
+	 * getting out of RAM on some badly mispredicted read patterns.
 	 */
-	unsigned int nbytes = nblks << zf->zf_dnode->dn_datablkshift;
+	unsigned int dbs = zf->zf_dnode->dn_datablkshift;
+	unsigned int nbytes = nblks << dbs;
 	unsigned int pf_nblks;
 	if (fetch_data) {
 		if (unlikely(zs->zs_pf_dist < nbytes))
 			zs->zs_pf_dist = nbytes;
-		else if (zs->zs_pf_dist < zfetch_min_distance)
+		else if (zs->zs_pf_dist < zfetch_min_distance &&
+		    (zs->zs_pf_dist < (1 << dbs) ||
+		    aggsum_compare(&zfetch_sums.zfetchstat_io_active,
+		    arc_c_max >> (4 + dbs)) < 0))
```

Comment:
Given that we're tracking in flight blocks and that block size can vary, I don't see how this works out to 6% of the ARC.  What am I missing here?

Reply:
Shift by 4 additional bits means 6.25%.  But you are right, it is very approximate, based on assumptions that all blocks are more or less of the same size.  Any way we have no idea what is the compressed size of prefetched blocks, so for compressible data it is an upper bound.

I am open to better ideas.  I just don't don't want to drag the block size to dmu_zfetch_done() to account bytes.  And I don't want to set hard limit on number of blocks, since it may appear too low for large systems, while too high for small ones.  So I am trying to estimate number of blocks depending on ARC size and current block size.

Reply:
That makes sense.  Unfortunately, I don't have a better idea.  The important thing here is that there is some reasonable limit is being enforced which scales with the ARC size.  This seems like a reasonable way to handle it.  Although I wonder if we should to provide a tunable or a kstat counter to indicate how often we're bumping up against the limit.

Reply:
I hope this limit to not be reached often, because it likely means our prefetcher generates too much extra traffic.  At this point we have only one report of this issue that I know about, so hopefully it is not so widespread.